Spigot-Server-Patches/0553-NamespacedKey-based-plugin-chunk-tickets.patch

```diff
@@ -0,0 +1,267 @@
++    }
++
++    @Override
++    public void removeKeyedPluginChunkTickets(org.bukkit.NamespacedKey key) {
++        Preconditions.checkArgument(key != null, "null key");
++
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        chunkDistanceManager.removeIfAllTicketsFor(TicketType.KEYED_PLUGIN_TICKET,
++                ticket -> key.equals(ticket.getObjectReason()));
++    }
++
++    @Override
++    public void removeKeyedPluginChunkTickets(Plugin plugin) {
++        Preconditions.checkArgument(plugin != null, "null plugin");
++
++        String namespace = new org.bukkit.NamespacedKey(plugin, "ignored").getNamespace();
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        chunkDistanceManager.removeIfAllTicketsFor(TicketType.KEYED_PLUGIN_TICKET,
++                ticket -> namespace.equals(ticket.getObjectReason().getNamespace()));
++    }
++
++    @Override
++    public Collection<io.papermc.paper.TickingLevel> getKeyedPluginChunkTicketLevels(int x, int z, org.bukkit.NamespacedKey key) {
++        Preconditions.checkArgument(key != null, "null key");
++
++        // logic copied from getPluginChunkTickets
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        Collection<Ticket<?>> tickets = chunkDistanceManager.tickets.get(ChunkCoordIntPair.pair(x, z));
++        if (tickets == null) {
++            return Collections.emptyList();
++        }
++
++        ImmutableList.Builder<io.papermc.paper.TickingLevel> ret = ImmutableList.builder();
++        for (Ticket<?> ticket : tickets) {
++            if (ticket.getTicketType() == TicketType.KEYED_PLUGIN_TICKET && key.equals(ticket.getObjectReason())) {
++                ret.add(io.papermc.paper.PaperTickingLevel.getTickingLevel(ticket.getTicketLevel()));
++            }
++        }
++        return ret.build();
++    }
++
++    @Override
++    public Collection<org.bukkit.NamespacedKey> getKeyedPluginChunkTickets(int x, int z) {
++        // logic copied from getPluginChunkTickets
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        Collection<Ticket<?>> tickets = chunkDistanceManager.tickets.get(ChunkCoordIntPair.pair(x, z));
++        if (tickets == null) {
++            return Collections.emptyList();
++        }
++
++        ImmutableList.Builder<org.bukkit.NamespacedKey> ret = ImmutableList.builder();
++        for (Ticket<?> ticket : tickets) {
++            if (ticket.getTicketType() == TicketType.KEYED_PLUGIN_TICKET) {
++                ret.add((org.bukkit.NamespacedKey) ticket.getObjectReason());
++            }
++        }
++        return ret.build();
++    }
++
++    @Override
++    public Map<org.bukkit.NamespacedKey, Collection<Chunk>> getKeyedPluginChunkTickets() {
```

Comment:
This really is better suited as a ListMultimap. I know spigot used this pattern for the other, but this is what ArrayListMultimap is for and we should just use the better collection.

Reply:
immutable result is not really important. users will learn quickly that mutating it does nothing.

Reply:
We also really should have a more useful API:

Set<Chunk> getKeyedPluginChunkTickets(NamespacedKey key)

This would let plugins tag chunks by an identifier then find the chunks by that identifier.